lib/galaxy/datatypes/data.py

```diff
@@ -1205,6 +1205,143 @@ def regex_line_dataprovider(
 class Directory(Data):
     """Class representing a directory of files."""
 
+    # The behavior of this class is intended to be similar to that of
+    # composite types, but with arbitrary structure of extra_files.
+
+    # A prioritized list of files in the directory that, if present, can serve
+    # as a replacement for a composite type's primary dataset.
+    recognized_index_files = []
+
+    # Directories converted from archives, and possibly others, have a single
+    # root folder inside the directory. The root_folder attribute lets tools
+    # access that folder without first exploring the directory tree themselves.
+    # Will be set to the empty string for flat directories.
+    MetadataElement(
+        name="root_folder",
+        default=None,
+        desc="Name of the root folder of the directory",
+        readonly=True,
+        optional=False,
+        visible=False,
+    )
+
+    # Capture the actual index file obtained from the recognized_index_files
+    # list for a concrete directory.
+    MetadataElement(
+        name="index_file",
+        default=None,
+        desc="Name of the index file relative to the root folder",
+        readonly=True,
+        optional=False,
+        visible=False,
+    )
+
+    MetadataElement(
+        name="folder_size",
+        default=0,
+        desc="Total size of the folder in bytes",
+        readonly=True,
+        optional=False,
+        visible=False,
+    )
+
+    @classmethod
+    def _get_size(cls, root_path):
+        total_size = 0
+        for dirpath, dirnames, filenames in os.walk(root_path):
+            for f in filenames:
+                fp = os.path.join(dirpath, f)
+                # skip if it is symbolic link
```

Comment:
Should we allow links in directories at all?

Reply:
So, for directories created via the converter, links are checked for safety, and excluded if they point to higher up than the top-level folder of their archive.
If we had tools generating directories, then I don't think we currently have a protection mechanism, but would have to investigate the tool output manually?

Reply:
I don't think that's the responsibility of a datatuype, and I'm pretty sure that can't be moved into the object store, and if it can that would be the right place to check this.